futures-util/src/future/shared.rs

```diff
@@ -97,52 +123,73 @@ impl<F> Shared<F> where F: Future {
         }
     }
 
-    fn set_waiter(&mut self, cx: &mut task::Context) {
-        let mut waiters = self.inner.notifier.waiters.lock().unwrap();
-        waiters.insert(self.waiter, cx.waker().clone());
+    /// Registers the current task to receive a wakeup when `Inner` is awoken.
+    fn set_waker(&mut self, cx: &mut task::Context) {
+        let mut wakers = self.inner.notifier.wakers.lock().unwrap();
+        if self.waker_key == NULL_WAKER_KEY {
+            self.waker_key = wakers.insert(Some(cx.waker().clone()));
+        } else {
+            let waker_slot = &mut wakers[self.waker_key];
+            let needs_replacement = if let Some(old_waker) = waker_slot {
+                // If there's still an unwoken waker in the slot, only wake
+                // if the current one wouldn't wake the same task.
+                !old_waker.will_wake(cx.waker())
+            } else {
+                true
+            };
+            if needs_replacement {
+                *waker_slot = Some(cx.waker().clone());
+            }
+        }
+        debug_assert!(self.waker_key != NULL_WAKER_KEY);
     }
 
-    unsafe fn clone_result(&self) -> Result<SharedItem<F::Item>, SharedError<F::Error>> {
-        match *self.inner.result.get() {
-            Some(Ok(ref item)) => Ok(SharedItem { item: item.item.clone() }),
-            Some(Err(ref e)) => Err(SharedError { error: e.error.clone() }),
-            _ => unreachable!(),
+    /// Safety: callers must first ensure that `self.inner.state`
+    /// is `COMPLETE`
+    unsafe fn clone_result(&self) -> Arc<F::Output> {
+        if let FutureOrOutput::Output(item) = &*self.inner.future_or_output.get() {
+            item.clone()
+        } else {
+            unreachable!()
         }
     }
 
-    fn complete(&self) {
-        unsafe { *self.inner.future.get() = None };
+    /// Safety: polling must have completed and self.inner.result must be
+    /// FutureOrOutput::Output
+    unsafe fn complete(&self) {
```

Comment:
This function merely contains two lines of code and is called only in a single place => It should be inlined

Reply:
Done.